app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabWebContentsObserver.java

```diff
@@ -0,0 +1,80 @@
+package com.igalia.wolvic.browser.api.impl;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import com.igalia.wolvic.browser.api.WSession;
+
+import org.chromium.content_public.browser.LifecycleState;
+import org.chromium.content_public.browser.NavigationHandle;
+import org.chromium.content_public.browser.WebContents;
+import org.chromium.content_public.browser.WebContentsObserver;
+import org.chromium.url.GURL;
+
+public class TabWebContentsObserver extends WebContentsObserver {
+    private @NonNull WebContents mWebContents;
+    private @NonNull SessionImpl mSession;
+
+    public TabWebContentsObserver(WebContents webContents, @NonNull SessionImpl session) {
+        super(webContents);
+        mWebContents = webContents;
+        mSession = session;
+    }
+
+    @Override
+    public void didStartLoading(GURL url) {
+        @Nullable WSession.ProgressDelegate delegate = mSession.getProgressDelegate();
+        if (delegate != null) {
+            delegate.onPageStart(mSession, url.toString());
+        }
+        dispatchCanGoBackOrForward();
+    }
+
+    @Override
+    public void didRedirectNavigation(NavigationHandle navigationHandle) {
+        dispatchCanGoBackOrForward();
```

Comment:
Don't you need to update the progress delegate with the new URL? Maybe with a new onPageStart session?

Reply:
There is no suitable API like onPageRedirect in  WSession.ProgressDelegate and |url| is not used in all listeners of onPageStart. And I'm worried we will call multiple onPageStart since TelemetryService could be affected.

Reply:
Fair enough